tango/integrations/transformers/ia3.py

```diff
@@ -0,0 +1,277 @@
+
+
+        :param linear_layer:
+            A :class:`~torch.nn.Linear` layer to adapt.
+        :param ia3_param_names:
+            A `str` to use as the name of ia3 parameters.
+        :param unfuse_size:
+            An `int` indicating hidden dimension of the query, key, and value vectors.
+            To be used only when the layer to modify is a fused projection of query,
+            key, and value vectors in an attention mechanism.
+        """
+        assert unfuse_size is None or (linear_layer.out_features == unfuse_size * 3)
+        self.in_features = linear_layer.in_features
+        self.out_features = linear_layer.out_features
+        self.unfuse_size = unfuse_size
+
+        super().__init__(ia3_param_names, unfuse_size)
+
+        self.weight = linear_layer.weight
+        self.bias = linear_layer.bias
+
+    def forward(self, x):
+        x = F.linear(x, self.weight, self.bias)
+        return self.scale_by_ia3(x)
+
+
+class Conv1DWithIA3(WithIA3):
+    def __init__(self, conv1d_layer: Conv1D, ia3_param_names: str, unfuse_size: int = None):
+        """
+        A replacement for :class:`~transformers.modeling_utils.Conv1D` modified with an IA3 adaptor
+
+
+        :param conv1d_layer:
+            A :class:`~transformers.modeling_utils.Conv1D` layer to adapt.
+        :param ia3_param_names:
+            A `str` to use as the name of ia3 parameters.
+        :param unfuse_size:
+            An `int` indicating hidden dimension of the query, key, and value vectors.
+            To be used only when the layer to modify is a fused projection of query,
+            key, and value vectors in an attention mechanism.
+        """
+        assert unfuse_size is None or (conv1d_layer.nf == unfuse_size * 3)
+
+        # nf: number of output features; nx: number of input features
+        self.out_features = conv1d_layer.nf
+        self.unfuse_size = unfuse_size
+
+        super().__init__(ia3_param_names, unfuse_size)
+
+        self.weight = conv1d_layer.weight
+        self.bias = conv1d_layer.bias
+
+    def forward(self, x):
+        # copied and pasted from the original Conv1D implemnetation
+        size_out = x.size()[:-1] + (self.out_features,)
+        x = torch.addmm(self.bias, x.view(-1, x.size(-1)), self.weight)
+        x = x.view(size_out)  # ... * self.nf
+
+        return self.scale_by_ia3(x)
+
+
+def modify_with_ia3(
+    transformer: PreTrainedModel, config: WithIA3Config, *, only_ia3_requires_grad: bool = True
+) -> PreTrainedModel:
+    """
+    A function to add ia3 adaptors to the given transformer. Code modified from
+    `t-few <https://github.com/r-three/t-few/blob/217cfa3b73aa66a07594826e4ebbbc516b331461/src/models/lora.py>`_
+    and Qinyuan Ye
+
+
+    :param model:
+        A :class:`~transformers.PreTrainedModel` to modify.
+    :param config:
+        A :class:`~tango.integrations.transformers.ia3.WithIA3Config` that specifies the layers to modify.
```

Comment:
Is there any chance we could automatically detect the right config, at least in some cases?

Reply:
So we could make it look up the known configs in `MODEL_NAME_TO_CONFIG` in [tango/integrations/transformers/ia3.py](https://github.com/allenai/tango/pull/403/files/68db503f1da57b4c1f8498819c084158badb8830#diff-f3d77c73a5bc3a8ac400e155fd8bbcd531b7b93c71759fb487d22ac20ea10425).

But if you mean trying to figure out a config from scratch just based on looking at the model architecture, that might be pretty difficult. Just from the few that we support right now they all have very different names for the layers we need. And there's enough variation in how the nodes can be nested that we can't find the layers we need by just looking for a Linear layer with a certain position in the model graph.

Reply:
No, I mean, can we just find the name of the model given the model, and do the lookup that way?

Reply:
I checked: We can look up `transformer_model.config.name_or_path`. If it matches anything in the dictionary of configs, we can use it automatically.

Reply:
Yup working on a commit to do that!

Reply:
Okay it's working now and I also updated the Catwalk end to use this: https://github.com/allenai/catwalk/pull/81/commits/9800e1249f5e311b16429635bfa957da6f3fd184